agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -1146,7 +1146,25 @@ impl DeriveSchema for TaggedOperator {
                     set! {then_schema, else_schema},
                 )))
             }
-            TaggedOperator::Filter(_) => todo!(),
+            TaggedOperator::Filter(f) => {
+                let input_schema = f.input.derive_schema(state)?;
+                let is_nullable = input_schema.satisfies(&NULLISH.clone()) != Satisfaction::Not;
+                let array_schema = array_schema_or_error!(input_schema, f.input);
+                let var_name = f._as.clone().unwrap_or("this".to_string());
+                state.variables.insert(var_name.clone(), array_schema);
+                f.cond.derive_schema(state)?;
+                let filter_schema = state.variables.remove(&var_name);
+                if let Some(schema) = filter_schema {
```

Comment:
I really like the idea of inserting and removing the variable, but this is unfortunately (very rarely) incorrect. If that variable already exists in the state, we are losing it forever here. We, unfortunately, need to clone the state and revert to it rather than this much more efficient implementation... or we can just check for key collision and only restore that one key. Up to you.

Reply:
ahhhh good call, I'll handle this with the clone